File: typescript/tests/fitness_board.test.ts
// fitness_board.test.ts
// Verifies the CircleAI.Fitness port: weekly workouts, calories-since, goals,
// exercise sets.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryFitnessBoard,
  FitnessDomainContext,
  workout,
  fitnessGoal,
  exerciseSet,
} from "../src/fitness/index";

describe("InMemoryFitnessBoard", () => {
  it("lists this week's workouts oldest-first from the Sunday week-start", () => {
    const b = new InMemoryFitnessBoard();
    const now = new Date("2026-01-07T12:00:00Z"); // Wed; week start Sun 2026-01-04
    b.log(workout("w0", "u1", "run", 30, 300, new Date("2026-01-03T23:00:00Z"))); // before
    b.log(workout("w1", "u1", "run", 30, 300, new Date("2026-01-06T09:00:00Z")));
    b.log(workout("w2", "u1", "lift", 45, 200, new Date("2026-01-05T09:00:00Z")));
    assert.deepEqual(
      b.workoutsThisWeek("u1", now).map((w) => w.workoutId),
      ["w2", "w1"],
    );
  });

  it("sums calories since a cutoff for one user", () => {
    const b = new InMemoryFitnessBoard();
    b.log(workout("w1", "u1", "run", 30, 300, new Date("2026-01-01T09:00:00Z")));
    b.log(workout("w2", "u1", "run", 30, 250, new Date("2026-01-05T09:00:00Z")));
    b.log(workout("w3", "u2", "run", 30, 999, new Date("2026-01-05T09:00:00Z")));
    assert.equal(b.totalCaloriesSince("u1", new Date("2026-01-02T00:00:00Z")), 250);
  });

  it("stores goals per user and sets per workout", () => {
    const b = new InMemoryFitnessBoard();
    b.setGoal(fitnessGoal("g1", "u1", "weight", 75, new Date("2026-06-01T00:00:00Z")));
    b.setGoal(fitnessGoal("g2", "u2", "5k", 22, new Date("2026-06-01T00:00:00Z")));
    assert.deepEqual(
      b.goalsFor("u1").map((g) => g.goalId),
      ["g1"],
    );
    b.addSet(exerciseSet("s1", "w1", "squat", 5, 100));
    b.addSet(exerciseSet("s2", "w1", "bench", 5, 80));
    b.addSet(exerciseSet("s3", "w2", "row", 8, 60));
    assert.deepEqual(
      b.setsFor("w1").map((s) => s.setId),
      ["s1", "s2"],
    );
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(FitnessDomainContext.systemPromptSnippet.includes("[DOMAIN: Fitness]"));
    assert.deepEqual(FitnessDomainContext.complianceFlags, ["HPCSA_Fitness", "POPIA", "Not_Medical_Advice"]);
    assert.deepEqual(FitnessDomainContext.suggestedTools, [
      "fitness_tracker",
      "exercise_db",
      "nutrition_tools",
      "analytics",
    ]);
  });
});

File: go/tests/fitness_board_test.go
// fitness_board_test.go
//
// Verifies the CircleAI.Fitness port (fitness_board.go): this-week workouts,
// calories-since totals, goals-for filtering (sorted), and sets-for insertion
// order.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestFitness_WorkoutsAndCalories(t *testing.T) {
	b := circleai.NewInMemoryFitnessBoard()
	now := time.Date(2026, 7, 8, 12, 0, 0, 0, time.UTC)
	b.Log(circleai.Workout{WorkoutId: "w1", UserId: "u1", Kind: "run", DurationMinutes: 30, CaloriesBurned: 300, AtUtc: now})
	b.Log(circleai.Workout{WorkoutId: "w2", UserId: "u1", Kind: "bike", DurationMinutes: 45, CaloriesBurned: 400, AtUtc: now.Add(-24 * time.Hour)})
	b.Log(circleai.Workout{WorkoutId: "w3", UserId: "u1", Kind: "swim", DurationMinutes: 20, CaloriesBurned: 200, AtUtc: now.Add(-14 * 24 * time.Hour)})

	wk := b.WorkoutsThisWeek("u1", now)
	if len(wk) != 2 || wk[0].WorkoutId != "w2" || wk[1].WorkoutId != "w1" {
		t.Fatalf("this-week workouts oldest-first failed: %+v", wk)
	}
	if cal := b.TotalCaloriesSince("u1", now.Add(-48*time.Hour)); cal != 700 {
		t.Fatalf("calories since = %v, want 700", cal)
	}
}

func TestFitness_GoalsAndSets(t *testing.T) {
	b := circleai.NewInMemoryFitnessBoard()
	b.SetGoal(circleai.FitnessGoal{GoalId: "g2", UserId: "u1", Metric: "weight", Target: 80})
	b.SetGoal(circleai.FitnessGoal{GoalId: "g1", UserId: "u1", Metric: "5k", Target: 25})
	b.SetGoal(circleai.FitnessGoal{GoalId: "g3", UserId: "u2", Metric: "steps", Target: 10000})

	goals := b.GoalsFor("u1")
	if len(goals) != 2 || goals[0].GoalId != "g1" || goals[1].GoalId != "g2" {
		t.Fatalf("goals-for sorted failed: %+v", goals)
	}

	b.AddSet(circleai.ExerciseSet{SetId: "s1", WorkoutId: "w1", Exercise: "squat", Reps: 5, WeightKg: 100})
	b.AddSet(circleai.ExerciseSet{SetId: "s2", WorkoutId: "w1", Exercise: "squat", Reps: 5, WeightKg: 105})
	b.AddSet(circleai.ExerciseSet{SetId: "s3", WorkoutId: "w2", Exercise: "bench", Reps: 8, WeightKg: 60})

	sets := b.SetsFor("w1")
	if len(sets) != 2 || sets[0].SetId != "s1" || sets[1].SetId != "s2" {
		t.Fatalf("sets-for insertion order failed: %+v", sets)
	}
}
